### ISSUE TITLE
Hooks for case changing commands

### ISSUE THREAD

**Issue by moewew** [latex3/latex3#1108]
`biblatex` had (https://github.com/plk/biblatex/pull/1233) some code to hook into `\MakeUppercase`/`\MakeLowercase` to change some of its (protected) commands to output the expected casing.

https://github.com/plk/biblatex/blob/8bd02c1502d9a4677e4554140740a6bd95847137/tex/latex/biblatex/biblatex.sty#L5058-L5117

The idea was that basically `biblatex` would take care of doing the case changing for these commands. Much like the following

```latex
\documentclass{article}

\DeclareRobustCommand*{\normalfoo}{Foo}
\DeclareRobustCommand*{\lcfoo}{foo}
\DeclareRobustCommand*{\ucfoo}{FOO}

\begin{document}
\normalfoo % -> Foo

% the idea is that \MakeLowercase does
% "\let\normalfoo\lcfoo"
\MakeLowercase{\normalfoo} % \lcfoo -> foo


% the idea is that \MakeUppercase does
% "\let\normalfoo\ucfoo"
\MakeUppercase{\normalfoo} % \ucfoo -> FOO
\end{document}
```

In the example things could be made expandable and the case changer could deal with them, but in `biblatex` some commands are much more complicated and cannot be easily made expandable.

Would it be possible to get some initialisation hook depending on the expected output (lowercase/uppercase) that can take care of capitalising protected stuff "manually".

See also https://github.com/plk/biblatex/issues/1234.

**Comment by josephwright** [comment]
I think we want a way for this to continue to work inside the current `expl3` code rather than just in `\MakeUppercase`. That suggests we need a 'substitution mechanism' plus a command that is 'case-mapping aware'. Based on @u-fischer's suggestion, I was thinking of doing that using

- `\text_case_select:nnnn <unchange> <upper> <lower> <title>`, which would be a way to 'force' a particular output
- `\text_declare_case_equivalent:Nn`, similar to the expansion one

That would allow something like
```latex
\text_declare_case_equivalent:Nn \bibstring \blx@bibtring@in@case 
\def\blx@bibtring@in@case{\text_case_select:nnnn{\bibstring}{\bibstring@upper}{\bibstring@lower}{\bibstring@upper}}
```
which would then mean that you'd get the same ability as you have now.

**Comment by moewew** [comment]
Is `<title>` only going to be used at the beginning of the title-cased string (so only for the capitalised part) or for all occurrences?

**Comment by josephwright** [comment]
@moewew  It would only be for the first 'character', after that we switch to either do-nothing or lowercasing.

**Comment by josephwright** [comment]
A sketch:
```latex
\documentclass[ngerman]{article}
\usepackage[T1]{fontenc}
\usepackage{babel}
\usepackage{csquotes}
\usepackage[style=authoryear]{biblatex}

\ExplSyntaxOn
\cs_new:Npn \text_case_select:nnnn #1#2#3#4
  {
    \__text_case_select_marker:
    #1
  }
\cs_new:Npn \__text_case_select_marker: { }

\tl_put_right:Nn \l_text_expand_exclude_tl
  { \text_case_select:nnnn }

\cs_new:Npn \__text_change_case_select:nnN #1#2#3
  {
    \cs_if_eq:NNTF #3 \text_case_select:nnnn
      { \use:c { __text_change_case_select_ #1 :nnNnnnn  } }
      { \use:c { __text_change_case_letterlike_ #1 :nnN } }
        {#1} {#2} #3
  }
\cs_new:Npn \__text_change_case_select_lower:nnNnnnn #1#2#3#4#5#6#7
  {
    \__text_change_case_store:n {#5}
    \__text_change_case_loop:nnw {#1} {#2}
  }
\cs_new:Npn \__text_change_case_select_upper:nnNnnnn #1#2#3#4#5#6#7
  {
    \__text_change_case_store:n {#6}
    \__text_change_case_loop:nnw {#1} {#2}
  }
\cs_new:Npn \__text_change_case_select_title:nnNnnnn #1#2#3#4#5#6#7
  {
    \__text_change_case_store:n {#7}
    \__text_change_case_loop:nnw {#1} {#2}
  }

\cs_gset:Npn \__text_change_case_exclude:nnNN #1#2#3#4
  {
    \__text_if_recursion_tail_stop_do:Nn #4
      { \__text_change_case_replace:nnN {#1} {#2} #3 }
    \str_if_eq:nnTF {#3} {#4}
      {
        \__text_use_i_delimit_by_q_recursion_stop:nw
          { \__text_change_case_exclude:nnNw {#1} {#2} #3 }
      }
      { \__text_change_case_exclude:nnNN {#1} {#2} #3 }
  }
\cs_new:Npn \__text_change_case_replace:nnN #1#2#3
  {
    \cs_if_exist:cTF { l__text_case_ \token_to_str:N #3 _tl }
      {
        \exp_args:Nv \__text_change_case_replace:nnn
          { l__text_case_ \token_to_str:N #3 _tl } {#1} {#2}
      }
      { \__text_change_case_select:nnN {#1} {#2} #3 }
  }
\cs_new:Npn \__text_change_case_replace:nnn #1#2#3
  { \__text_change_case_loop:nnw {#2} {#3} #1 }


\cs_new_protected:Npn \text_declare_case_equivalent:Nn #1#2
  {
    \tl_clear_new:c { l__text_case_ \token_to_str:N #1 _tl }
    \tl_set:cn { l__text_case_ \token_to_str:N #1 _tl } {#2}
  }
  
\makeatletter
\text_declare_case_equivalent:Nn \bibstring
{
 \text_case_select:nnnn
    \bibstring
    \blx@imc@bibucstring
    \blx@imc@biblcstring
    \blx@imc@bibucstring
}
\makeatother

\ExplSyntaxOff

% \blx@imc@bibucstring

\addbibresource{biblatex-examples.bib}

\renewbibmacro{cite}{%
  \bibstring{bibliography}%
  \setunit{\addspace}%
  {
%  \tracingall
  \MakeUppercase{\bibstring{bibliography}}%
  }
  \setunit{\addspace}%
  \MakeLowercase{\bibstring{bibliography}}}

\begin{document}
\cite{sigfridsson}

\end{document}

  
 
\begin{filecontents}[overwrite]{testpackage.sty}
\NeedsTeXFormat{LaTeX2e}[2022-06-01]
\ProvidesExplPackage
  {testpackage}
  {2022/07/02} {}
  {A package for testing}

\keys_define:nn { testpackage }
  {
    , unknown .code:n = { \showtokens{\CurrentOption} \PassOptionsToPackage {#1} { create-theorem } }
  }
\ProcessKeyOptions [ testpackage ]

\endinput
\end{filecontents}
\documentclass{article}

\usepackage[originalref]{testpackage}
\usepackage{create-theorem}

\begin{document}

\ExplSyntaxOn

\bool_if:NTF \l__crthm_regionalref_bool
  { Option ~ does ~ not ~ work }
  { Option ~ works }

\ExplSyntaxOff

\end{document}
```

**Comment by josephwright** [comment]
The idea here is to provide two parts to the mechanism, a way to swap out commands in case changing, and an equivalent of `\texorpdfstring` for case changing. I think that covers the use case and applies more widely - thoughts? (@u-fischer, @davidcarlisle)

**Comment by josephwright** [comment]
We'd likely want a 2e name here too ...

**Comment by u-fischer** [comment]
The interface looks fine, also if one use it on the fly. Regarding the 2e name `\CaseSwitch` or `\UseCaseSwitch`   ?

~~~~
\documentclass{article}
\usepackage[T1]{fontenc}

\ExplSyntaxOn
\cs_new:Npn \text_case_select:nnnn #1#2#3#4
  {
    \__text_case_select_marker:
    #1
  }
\cs_new:Npn \__text_case_select_marker: { }

\tl_put_right:Nn \l_text_expand_exclude_tl
  { \text_case_select:nnnn }

\cs_new:Npn \__text_change_case_select:nnN #1#2#3
  {
    \cs_if_eq:NNTF #3 \text_case_select:nnnn
      { \use:c { __text_change_case_select_ #1 :nnNnnnn  } }
      { \use:c { __text_change_case_letterlike_ #1 :nnN } }
        {#1} {#2} #3
  }
\cs_new:Npn \__text_change_case_select_lower:nnNnnnn #1#2#3#4#5#6#7
  {
    \__text_change_case_store:n {#5}
    \__text_change_case_loop:nnw {#1} {#2}
  }
\cs_new:Npn \__text_change_case_select_upper:nnNnnnn #1#2#3#4#5#6#7
  {
    \__text_change_case_store:n {#6}
    \__text_change_case_loop:nnw {#1} {#2}
  }
\cs_new:Npn \__text_change_case_select_title:nnNnnnn #1#2#3#4#5#6#7
  {
    \__text_change_case_store:n {#7}
    \__text_change_case_loop:nnw {#1} {#2}
  }

\cs_gset:Npn \__text_change_case_exclude:nnNN #1#2#3#4
  {
    \__text_if_recursion_tail_stop_do:Nn #4
      { \__text_change_case_replace:nnN {#1} {#2} #3 }
    \str_if_eq:nnTF {#3} {#4}
      {
        \__text_use_i_delimit_by_q_recursion_stop:nw
          { \__text_change_case_exclude:nnNw {#1} {#2} #3 }
      }
      { \__text_change_case_exclude:nnNN {#1} {#2} #3 }
  }
\cs_new:Npn \__text_change_case_replace:nnN #1#2#3
  {
    \cs_if_exist:cTF { l__text_case_ \token_to_str:N #3 _tl }
      {
        \exp_args:Nv \__text_change_case_replace:nnn
          { l__text_case_ \token_to_str:N #3 _tl } {#1} {#2}
      }
      { \__text_change_case_select:nnN {#1} {#2} #3 }
  }
\cs_new:Npn \__text_change_case_replace:nnn #1#2#3
  { \__text_change_case_loop:nnw {#2} {#3} #1 }


\cs_new_protected:Npn \text_declare_case_equivalent:Nn #1#2
  {
    \tl_clear_new:c { l__text_case_ \token_to_str:N #1 _tl }
    \tl_set:cn { l__text_case_ \token_to_str:N #1 _tl } {#2}
  }

%for testing  
\cs_set_eq:NN\CaseSwitch\text_case_select:nnnn  
\cs_set_eq:NN\MakeTitlecase\text_titlecase:n
\tl_put_right:Nn \l_text_expand_exclude_tl
  { \CaseSwitch }
\ExplSyntaxOff

\parskip=10pt \parindent=0pt

\begin{document}

\CaseSwitch{normal}{lower}{UPPER}{Title}

\MakeUppercase{some word foo}\\
\MakeUppercase{\CaseSwitch{normal}{lower}{UPPER}{Title}}\\
\MakeUppercase{some word \CaseSwitch{normal}{lower}{UPPER}{Title}}


\MakeLowercase{SOME WORD FOO}\\
\MakeLowercase{\CaseSwitch{normal}{lower}{UPPER}{Title}}\\
\MakeLowercase{SOME WORD \CaseSwitch{normal}{lower}{UPPER}{Title}}\\


\MakeTitlecase{some word foo}\\
\MakeTitlecase{\CaseSwitch{normal}{lower}{UPPER}{Title}}\\
\MakeTitlecase{some word \CaseSwitch{normal}{lower}{UPPER}{Title}} 
\end{document}
~~~~


![image](https://user-images.githubusercontent.com/4047173/177137286-7076d56f-5186-4b78-a047-a27ee0f85542.png)

**Comment by josephwright** [comment]
@u-fischer  You fancy adding `\MakeTitlecase` while we are at it?

**Comment by josephwright** [comment]
Does the `expl3` naming look OK? I see @u-fischer used 'switch', whereas I went for 'select' - perhaps `\text_case_switch:nnnn` would be better (and a match).

**Comment by davidcarlisle** [comment]
> @u-fischer You fancy adding `\MakeTitlecase` while we are at it?

I nearly suggested that when doing the `\MakeUppercase` update, would make sense.

At some point (not this PL) we may want `\MakeXxxxcase[abc]{....}` wrapping the `:nn` versions of the l3 text code as well.

**Comment by josephwright** [comment]
> > @u-fischer You fancy adding `\MakeTitlecase` while we are at it?
> 
> I nearly suggested that when doing the `\MakeUppercase` update, would make sense.

Now or for the Fall 2022 release?

> At some point (not this PL) we may want `\MakeXxxxcase[abc]{....}` wrapping the `:nn` versions of the l3 text code as well.

Yes, I think that's one for the Fall 2022 release

**Comment by josephwright** [comment]
I get the feeling my `expl3` code looks OK - unless I hear otherwise I'll add a bit later today

**Comment by davidcarlisle** [comment]
`switch` better than `select` I think. Not sure why, possibly because the selection key is implicit: `select 2 {a,b,c,d}` selecting `b`  but here  you are not specifying a selection, rather offering choices to be selected later.

**Comment by davidcarlisle** [comment]
> Now or for the Fall 2022 release?

Now I think, otherwise it is harder to document the 4th argument  of the case switch in a 2e context

**Comment by josephwright** [comment]
Fixed by 66838cc7a74fa291538b2ab0d789dc94d8be2b9b at the `expl3` end
